text/3334-niche.md

```diff
@@ -0,0 +1,482 @@
+  range *to* the compiler.
+- **Invalid-pointer niches**: On targets where certain pointer values cannot
+  represent a valid pointer in a given context (such as on x86-64 where the
+  upper half of the address space represents kernel-space address and the lower
+  half represents userspace addresses), types containing such pointers could use
+  a large swathe of values as a niche.
+- **Pointer high-bit niches**: On targets that don't permit addresses with some of
+  the high bits set (such as implicitly on historical x86 or ARM platforms, or
+  explicitly defined via ARM's "top-byte ignore" or AMD's "upper address
+  ignore" or Intel's "Linear Address Masking"), types containing pointers could
+  potentially use values with those high bits set as a niche. This would likely
+  require compile-time configuration.
+- **Multiple niches**: A type could define multiple niches, rather than just a
+  single range.
+- **Other bit-pattern niches**: A type could define niches via a bit pattern,
+  rather than a range.
+- **Per-field niches**: A struct containing multiple fields could have a niche
+  on a specific field, rather than the whole struct.
+- **structs with ZST fields**: A struct could contain fields with zero-sized
+  types (e.g. `PhantomData`) and still have a niche.
+- **Fields of reference type**: In addition to allowing raw pointers, structs
+  with niches could allow references. In practice, if the references have a
+  lifetime other than `'static`, this will also require at least some support
+  for generic parameters.
+- **Non-primitive fields**: A struct could contain fields of non-primitive
+  types, such as tuples, arrays, or other structs (including structs with
+  niches themselves). This should wait until after niches support providing
+  values with the type of the field, rather than as an unsigned integer.
+- **Whole-struct niches**: A struct containing multiple non-zero-sized fields
+  could have niche values for the whole struct.
+- **Union niches**: A union could have a niche.
+- **Enum niches**: An enum or an enum variant could have a niche.
+- **Specified mappings into niches**: Users may want to rely on mappings of
+  multiple values into a multi-value niche. For instance, users could define a
+  type with a niche containing a range of integer values, and a range of
+  integer error codes, and rely on `Result<T, E>` assigning specific niche
+  values to specific error codes, in order to match a specific ABI (such as the
+  Linux kernel's `ERR_PTR`).
+- **Safety**: The attribute specified in this RFC requires an unsafe block to
+  set the field. Future extensions could allow safely setting the field, after
+  verifying in a compiler-visible manner that the value does not fall within
+  the niche. For instance, via `derive(TryFrom)` (see below), or by checking
+  the value of a compile-time constant expression to ensure that it does not
+  fall within the niche.
```

Comment:
A pattern-type mechanism for safe niches would allow safe assignment to fields of niched types at runtime. So if unsafe assertion that a value meets a niche uses the exact same syntax, inside a large `unsafe` block it could be very easy to accidentally perform an unsafe operation when you didn't mean to. `mystruct.field = myvalue;` looks like a normal assignment, acts like a normal assignment, and to know that it's actually a special `unsafe` operation you would need to inspect the (potentially inferred) type of `myvalue`. Ideally, setting a niched field with a value Rust doesn't trust should require a separate unsafe intrinsic, something like `transmute` (it could even be `transmute`).